Add Jetpack branding to Scan screens starting from Phase One

ovitrif opened the pull request:
Fixes Partially #17334

This PR adds the logic to display the Jetpack powered branding banner on the Scan screens starting from Phase One of features removal.

## To test:

Jetpack branding on Scan screens should be visible starting from phase one of features removal, while it should be hidden when all Jetpack removal flags are disabled.

Normal Phase - Jetpack Branding not shown
- This is the phase when the `Jetpack powered` should **not** be displayed
- Enable this phase by making sure all Jetpack removal flags are disabled

Branded Phase - Jetpack Branding shown
- This is the phase where we show a `Jetpack powered` branding on the Stats screens
- Enable this phase by making sure `jp_removal_one` flag is enabled
- Go to `Me` → `App Settings` → `Debug settings` and **enable** `jp_removal_one`
- Additionally, any other Jetpack removal phase flag should end up in the branding being shown 

Test the following steps in each phase:

#### Testing Steps

1. Open WordPress app and login with site address, ❗ the site should have Jetpack Scan _Daily_ enabled
   1. Additionally, login with your WP.com account and make sure Jetpack is connected to your WP.com account
   2. Make sure the selected site is the self-hosted one, from step `1`
2. Toggle one of the phases based on the notes above testing steps
3. Go to `My Site` → `Menu` tab
4. From menu, under `Jetpack` tap `Scan`
5. If in normal phase, expect no `Jetpack powered` branding
6. If in branded phase:
   1. **Expect** the `Jetpack powered` banner to be visible at the bottom of the screen
   2. Tap the `Jetpack powered` banner
   3. **Expect** the `WordPress is better with Jetpack` bottom sheet to be displayed
   4. **Optionally** tap the `Try the new Jetpack app` button
      1. If Jetpack app is not installed it should open the Play Store to install the app
      2. If Jetpack app is installed, it should be opened
   5. Go back to the WordPress app
7. Tap `History` at the top right of the Scan screen
8. If in normal phase, expect no `Jetpack powered` branding
9. If in branded phase:
   1. **Expect** the `Jetpack powered` banner to be visible at the bottom of the screen
   2. Tap the `Jetpack powered` banner
   3. **Expect** the `WordPress is better with Jetpack` bottom sheet to be displayed
   4. **Optionally** tap the `Try the new Jetpack app` button
      1. If Jetpack app is not installed it should open the Play Store to install the app
      2. If Jetpack app is installed, it should be opened

### Screenshots

> **Note** I've used dummy data for the screenshots to make the Scan screens scrollable.

| Jetpack Scan | Jetpack Scan History |
| - | - |
| <img width="314" src="https://user-images.githubusercontent.com/4588074/207313857-11af481c-cf22-43ca-900e-9cb9d963cdee.png"> | <img width="314" src="https://user-images.githubusercontent.com/4588074/207313881-aa276063-908c-4879-97a9-9e8b3d71d73e.png"> |

## Regression Notes
1. Potential unintended areas of impact
   Scan and Scan history screen.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing

3. What automated tests I added (or what prevented me from doing so)
   - Added tests in `JetpackFeatureRemovalBrandingUtilTest`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Add Jetpack powered banner to Scan screen

ovitrif committed:
Show Scan jetpack branding starting from phase one

ovitrif committed:
Add tests for JetpackFeatureRemovalBrandingUtil

ovitrif committed:
Add jetpack powered banner to Scan history

ovitrif committed:
Update test names for phase one branding

AjeshRPai commented on WordPress/src/main/java/org/wordpress/android/ui/jetpackoverlay/JetpackFeatureRemovalBrandingUtil.kt:
```diff
@@ -0,0 +1,24 @@
+    private val jetpackFeatureRemovalPhaseHelper: JetpackFeatureRemovalPhaseHelper,
+) {
+    fun shouldShowPhaseOneBranding(): Boolean {
+        if (buildConfigWrapper.isJetpackApp) return false
```
❓  Is the check `if (buildConfigWrapper.isJetpackApp) return false` necessary as `JetpackBrandingUtils.shouldShowJetpackBranding()` already has this check?

OR alternatively can we omit the usage of the class `JetpackBrandingUtils` and use the below function in `JetpackFeatureRemovalBrandingUtil`?. That way, we can call the `JetpackFeatureRemovalBrandingUtil` from Scan screens to check whether the branding needs to be displayed.



```
fun shouldShowJetpackBranding(): Boolean {
        return isWpComSite() && jetpackPoweredFeatureConfig.isEnabled() && !buildConfigWrapper.isJetpackApp
    }
```
Let me know your thoughts 🤔

ovitrif replied:
> ❓ Is the check if (buildConfigWrapper.isJetpackApp) return false necessary as JetpackBrandingUtils.shouldShowJetpackBranding() already has this check?

Thank you for raising the question @AjeshRPai 🙇 
I agree the check is redundant given it's already covered by `JetpackBrandingUtils`. I will remove it and adapt the unit tests 👍 

> OR alternatively can we omit the usage of the class `JetpackBrandingUtils` and use the below function in `JetpackFeatureRemovalBrandingUtil`?. That way, we can call the `JetpackFeatureRemovalBrandingUtil` from Scan screens to check whether the branding needs to be displayed.
> [...]
> Let me know your thoughts 🤔

My initial idea was to rely only on `JetpackBrandingUtils` in the UI layer since it's also used to update the visibility of the banner when scrolling and it also already had the logic for the additional checks. IIRC porting these checks to `JetpackFeatureRemovalBrandingUtil` would mean reimplementing them there which is something I've tried to avoid in the first place.
If you're ok with it my thoughts are to keep the current logic as is, without the redundant check for Jetpack app.

AjeshRPai replied:
Hey @ovitrif 

>IIRC porting these checks to JetpackFeatureRemovalBrandingUtil would mean reimplementing them there which is something I've tried to avoid in the first place.

Oh, I see. Then it makes sense to keep the current logic as is after removing the redundant check 👍🏼

AjeshRPai committed:
Adds: a missing test for phase not started

AjeshRPai commented on WordPress/src/test/java/org/wordpress/android/ui/jetpackoverlay/JetpackFeatureRemovalBrandingUtilTest.kt:
```diff
@@ -0,0 +1,49 @@
+
+        assertTrue(shouldShowBranding)
+    }
+}
```
I have added a test for a condition in which phase one is not started with the commit cb26c3af190213043c7f8f584c311bf5f8e58e36. I hope it's ok. Let me know if u want me to revert it.

ovitrif replied:
Thank you for adding this test, looks good to me and it's an important part in the logic, great to have it covered 👍

AjeshRPai reviewed (approved):
Hey @ovitrif 

I have tested the changes using a site having Scan daily and also have gone through the code changes. Everything looks good to me 👍🏼 🟢 

Tested all these scenario
- Jetpack-powered branding is not shown when the flag is disabled 🟢 
- Jetpack powered branding is shown in Scan and Scan history screen when jp removal flag 1 - 4 is enabled 🟢  
- Clicking on Jetpack powered badge opens the Jetpack-powered bottom sheet 🟢 
- Clicking on Try the new jetpack app redirects to the play store/jetpack app 
- Verified that the unit tests in JetpackFeatureRemovalBrandingUtilTest.kt pass and are accurate 🟢 

I have left two minor suggestions. I am approving but not merging the PR so that you can have a look at the suggestions. 

💡 Suggestion: Do you think it makes sense to migrate the DI in Fragments and Viewmodel related to Scan to Hilt?

You have added the Hilt entry points to the `ScanHistoryActivity` and `ScanActivity`. Hence I thought it would be best to migrate the associated components too. Feel free to take it up later as well.

ovitrif commented:
> I have tested the changes using a site having Scan daily and also have gone through the code changes. Everything looks good to me 👍🏼 🟢

Thank you for testing and validating the PR @AjeshRPai 🙇 

> 💡 Suggestion: Do you think it makes sense to migrate the DI in Fragments and Viewmodel related to Scan to Hilt?

Good idea, I'll work on migrating the DI in the updated fragments and their associated ViewModels in this PR 👍

ovitrif committed:
Migrate scan fragment and vm DI to Hilt

ovitrif committed:
Migrate scan history fragment DI to Hilt

ovitrif committed:
Migrate scan history view model DI to Hilt

We're retrieving the ViewModel of the parent fragment (`ScanHistoryFragment`) in `ScanHistoryListFragment`, in order to make sure we're using the same instance I've tested this with logging the hashCode, and confirmed the hashcode is the same.

ovitrif committed:
Remove redundant check for jetpack app

ovitrif commented:
@AjeshRPai 👋 Thank you for your suggestions 👍 

1️⃣ I've updated the PR to migrate the DI to Hilt of:
- Scan and ScanHistory fragments
- Scan and ScanHistory view models

Since `ScanHistoryViewModel` is also retrieved in the child fragments (`ScanHistoryListFragment`) I've checked to make sure the same instance is shared by verifying the viewModel has the same `hashCode()`

Given this change poses the risk of introducing a crash on Scan / Scan History I suggest to retest the PR changes 👍 

2️⃣ As for the [other suggestion](https://github.com/wordpress-mobile/WordPress-Android/pull/17668#discussion_r1048163437) I already went forward with removing the redundant check on `isJetpackApp` and if you agree [my proposal is to leave the current check logic as is](https://github.com/wordpress-mobile/WordPress-Android/pull/17668#discussion_r1048466913), to avoid reimplementing the additional checks.

Let me know wdyt and whether the updates are valid and testing goes well 🙇

AjeshRPai committed:
Updates: ScanHistoryListFragment and ScanHistoryListViewModel to Hilt

Updates: the ScanHistoryListFragment and ScanHistoryListViewModel
to hilt
Updates: the viewmodel fetching logic to have single parent viewmodel

AjeshRPai commented:
Hey @ovitrif 

I have tested the changes again, everything looks good to me. 👍🏼 🟢 . Thanks for making the Hilt changes. 🙌🏼 

During the review I found that 
1. `ScanHistoryListFragment` and `ScanHistoryListViewModel` is still using the dagger - So I updated both classes to use Hilt 
2. Both `ScanHistoryListFragment` and `ScanHistoryFragment` use the same instance of ViewModel. Hence I have updated the logic in `ScanHistoryFragment` from 
`private val viewModel: ScanHistoryViewModel by viewModels()`   
to 
` private val viewModel: ScanHistoryViewModel by activityViewModels()`
 
ScanHistoryListFragment also fetches the ViewModel by `activityViewModels()` delegate. I did a round of manual testing to ensure there were no issues. 

I have done the above changes in the commit [366e234](https://github.com/wordpress-mobile/WordPress-Android/pull/17668/commits/366e234147d09c386bcfc012f600d60e0105f370). Please have a look at it and let me know if that's ok. 

Once again, thanks for making the Hilt changes. You 🪨

AjeshRPai reviewed (approved):
Everything Looks good to me

AjeshRPai committed:
Removes: redundant import

ovitrif commented:
> I have tested the changes again, everything looks good to me. 👍🏼 🟢 . Thanks for making the Hilt changes. 🙌🏼
> 
> During the review I found that
> 
> 1. `ScanHistoryListFragment` and `ScanHistoryListViewModel` is still using the dagger - So I updated both classes to use Hilt
> 2. Both `ScanHistoryListFragment` and `ScanHistoryFragment` use the same instance of ViewModel. Hence I have updated the logic in `ScanHistoryFragment` from
>    `private val viewModel: ScanHistoryViewModel by viewModels()`
>    to
>    ` private val viewModel: ScanHistoryViewModel by activityViewModels()`
> 
> ScanHistoryListFragment also fetches the ViewModel by `activityViewModels()` delegate. I did a round of manual testing to ensure there were no issues.
> 
> I have done the above changes in the commit [366e234](https://github.com/wordpress-mobile/WordPress-Android/pull/17668/commits/366e234147d09c386bcfc012f600d60e0105f370). Please have a look at it and let me know if that's ok.
> 
> Once again, thanks for making the Hilt changes. You 🪨

Thank you for adding these improvements @AjeshRPai 🙇 , you are awesome 🥇 🚀 🎉 

I've did another round of testing on my side and confirm everything works as expected, your changes look good and work great, the shared `ScanHistoryViewModel` instance is preserved across fragments 🙇 

This looks ready on my side too, once the CI checks pass this PR is ready to be merged 🎉
